man/humanize_number.3
.\"	$NetBSD: humanize_number.3,v 1.8 2008/04/30 13:10:50 martin Exp $
.\"
.\" Copyright (c) 1999, 2002, 2008 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Luke Mewburn and by Tomas Svensson.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd February 9, 2008
.Dt HUMANIZE_NUMBER 3
.Os
.Sh NAME
.Nm dehumanize_number ,
.Nm humanize_number
.Nd format a number into a human readable form and viceversa
.Sh LIBRARY
.ds str-Lb-libbsd Utility functions from BSD systems (libbsd, \-lbsd)
.Lb libbsd
.Sh SYNOPSIS
.In stdlib.h
.Ft int
.Fn dehumanize_number "const char *str" "int64_t *result"
.Ft int
.Fn humanize_number "char *buf" "size_t len" "int64_t number" "const char *suffix" "int scale" "int flags"
.Sh DESCRIPTION
The
.Fn humanize_number
function formats the signed 64 bit quantity given in
.Fa number
into
.Fa buffer .
A space and then
.Fa suffix
is appended to the end.
.Fa buffer
must be at least
.Fa len
bytes long.
.Pp
If the formatted number (including
.Fa suffix )
would be too long to fit into
.Fa buffer ,
then divide
.Fa number
by 1024 until it will.
In this case, prefix
.Fa suffix
with the appropriate SI designator.
.Pp
The prefixes are:
.Bl -column "Prefix" "Description" "Multiplier" -offset indent
.It Sy "Prefix" Ta Sy "Description" Ta Sy "Multiplier"
.It k	kilo	1024
.It M	mega	1048576
.It G	giga	1073741824
.It T	tera	1099511627776
.It P	peta	1125899906842624
.It E	exa	1152921504606846976
.El
.Pp
.Fa len
must be at least 4 plus the length of
.Fa suffix ,
in order to ensure a useful result is generated into
.Fa buffer .
To use a specific prefix, specify this as
.Fa scale
(Multiplier = 1024 ^ scale).
This can not be combined with any of the
.Fa scale
flags below.
.Pp
The following flags may be passed in
.Pa scale :
.Bl -tag -width Dv -offset indent
.It Dv HN_AUTOSCALE
Format the buffer using the lowest multiplier possible.
.It Dv HN_GETSCALE
Return the prefix index number (the number of times
.Fa number
must be divided to fit) instead of formatting it to the buffer.
.El
.Pp
The following flags may be passed in
.Pa flags :
.Bl -tag -width Dv -offset indent
.It Dv HN_DECIMAL
If the final result is less than 10, display it using one digit.
.It Dv HN_NOSPACE
Do not put a space between
.Fa number
and the prefix.
.It Dv HN_B
Use 'B' (bytes) as prefix if the original result does not have a prefix.
.It Dv HN_DIVISOR_1000
Divide
.Fa number
with 1000 instead of 1024.
.El
.Pp
The
.Fn dehumanize_number
function parses the string representing an integral value given in
.Fa str
and stores the numerical value in the integer pointed to by
.Fa result .
The provided string may hold one of the suffixes, which will be interpreted
and used to scale up its accompanying numerical value.
.Sh RETURN VALUES
.Fn humanize_number
returns the number of characters stored in
.Fa buffer
(excluding the terminating NUL) upon success, or \-1 upon failure.
If
.Dv HN_GETSCALE
is specified, the prefix index number will be returned instead.
.Pp
.Fn dehumanize_number
returns 0 if the string was parsed correctly.
A \-1 is returned to indicate failure and an error code is stored in
.Va errno .
.Sh ERRORS
.Fn dehumanize_number
will fail and no number will be stored in
.Fa result
if:
.Bl -tag -width Er
.It Bq Er EINVAL
The string in
.Fa str
was empty or carried an unknown suffix.
.It Bq Er ERANGE
The string in
.Fa str
represented a number that does not fit in
.Fa result .
.El
.Sh SEE ALSO
.Xr humanize_number 9
.Sh HISTORY
.Fn humanize_number
first appeared in
.Nx 2.0 .
.Pp
.Fn dehumanize_number
first appeared in
.\" FIXME: This should be in groff, but for now it avoids the warning.
.ds operating-system-NetBSD-5.0 5.0
.Nx 5.0 .


ChangeLog
2008-03-16  Guillem Jover  <guillem@debian.org>

	* include/vis.h: Remove UC Berkeley advertising clause, as per
	<ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.
	* src/vis.c: Likewise.
	* src/unvis.c: Likewise.
	* src/heapsort.c: Likewise.
	* src/linkaddr.c: Likewise.
	* man/fgetln.3: Likewise.

2007-12-18  Petr Salinger  <Petr.Salinger@seznam.cz>

	* include/bsd/queue.h: sync with current glibc variant on GNU/kFreeBSD

2006-11-08  Aurelien Jarno  <aurelien@aurel32.net>

	* src/link_addr.c: New file.
	* include/bsd/if_dl.h: Add link_addr.c declarations.
	* Makefile (LIB_SRCS): Add link_addr.c.
	* Versions: add link_addr and link_ntoa

2006-08-25  Petr Salinger  <Petr.Salinger@seznam.cz>

	* src/humanize_number.c: New file.
	* include/libutil.h: New file with humanize_number declarations.
	* Makefile (LIB_SRCS): Add humanize_number.c.
	* Versions: Add humanize_number.

2006-03-29  Robert Millan  <rmh@aybabtu.com>

	* src/heapsort.c: New file.
	* include/bsd/stdlib.h: Add heapsort declaration.
	* Makefile (LIB_SRCS): Add heapsort.c.
	* Versions: Add heapsort.

2006-03-19  Robert Millan  <rmh@aybabtu.com>

	* src/vis.c: New file.
	* src/unvis.c: Ditto.
	* include/vis.h: Add vis/unvis declarations.
	* Makefile (LIB_SRCS): Add vis.c/unvis.c.
	* Versions: Add vis/unvis functions.

2006-03-17  Robert Millan  <rmh@aybabtu.com>

	* src/progname.c: New file.
	* include/bsd/stdlib.h: Add {get,set}progname declarations.
	* Makefile (LIB_SRCS): Add 'progname.c'.
	* Versions: Add getprogname and setprogname.

2006-02-15  Guillem Jover  <guillem@debian.org>

	* include/bsd/cdefs.h: Disable __unused, it conflicts with a struct
	member in a Linux header.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/stdlib.h: Remove <sys/types.h> and <stdio.h>.
	* include/bsd/getopt.h: Protect header from multiple inclusion.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h: Remove commented out macros.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h (setproctitle, __dead2, __unused, __printflike)
	(__FBSDID): Moved to ...
	* include/bsd/cdefs.h: ... here.
	(__unused, __printflike) [__GNUC__]: Use proper __attribute__.
	* Makefile (LIB_INCLUDES): Add 'cdefs.h'.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* arc4random.c: Move to ...
	* src/arc4random.c: ... here.
	* bsd_getopt.c: Move to ...
	* src/bsd_getopt.c:: ... here.
	* err.c: Move to ...
	* src/err.c:: ... here.
	* fgetln.c: Move to ...
	* src/fgetln.c:: ... here.
	* fmtcheck.c: Move to ...
	* src/fmtcheck.c:: ... here.
	* inet_net_pton.c: Move to ...
	* src/inet_net_pton.c:: ... here.
	* strlcat.c: Move to ...
	* src/strlcat.c:: ... here.
	* strlcpy.c: Move to ...
	* src/strlcpy.c:: ... here.
	* md5c.c: Move to ...
	* src/md5c.c:: ... here.
	* md5.copyright: Move to ...
	* src/md5.copyright:: ... here.
	* Makefile (LIB_SRCS): Automatically prefix the src/ dir.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_SRCS): Line wrap.
	(LIB_INCLUDES): List only the files, automatically prefix the dir.
	(LIB_MANS): Likewise.

2006-02-13  Robert Millan  <rmh@aybabtu.com>

	* bsd_getopt.c: New file.
	* include/bsd/getopt.h: Likewise.
	* Makefile (LIB_SRCS): Add 'bsd_getopt.c'.
	(LIB_INCLUDES): Add 'include/bsd/getopt.h'.
	* Versions: Added bsd_getopt and optreset.

2006-02-10  Robert Millan  <rmh@aybabtu.com>

	* err.c: New file.
	* include/bsd/err.h: Likewise.
	* Makefile (LIB_SRCS): Add 'err.c'.
	(LIB_INCLUDES): Add 'include/bsd/err.h'.
	* Versions: Added errc, warnc, verrc and vwarn.

2005-12-19  Aurelien Jarno <aurel32@debian.org>

	* man/arc4random.3: New file.
	* man/fgetln.3: Likewise.
	* man/fmtcheck.3: Likewise.
	* man/strlcpy.3: Likewise.
	* fmtcheck.c:: Likewise.
	* include/stdlib.h: Added fmtcheck declaration.
	* Versions: Added fmtcheck.

2005-11-01  Guillem Jover  <guillem@debian.org>

	* ip_icmp.h: Move to ...
	* include/bsd/ip_icmp.h: ... here.
	* md5.h: Move to ...
	* include/bsd/md5.h: ... here.
	* bsd.h: Move to ...
	* include/bsd/bsd.h: ... here. Split contents ...
	* include/bsd/random.h: ... here ...
	* include/bsd/queue.h: ... here ...
	* include/bsd/string.h: ... and here.
	* md5c.c: Change md5.h include line.
	* Makefile (MK_CFLAGS): Include from the proper dir.

2005-08-03  Hector Garcia  <hector@debian.org>

	* Makefile: Fixed to place soft links instead in soname, and
	shared_so instead of copying files.
	* md5.h (MD5End, MD5File, MD5FileChunk, MD5Data): Commented out
	functions not available on 'md5c.c'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_INCLUDES): Add 'md5.h'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Versions (LIBBSD_0.0): Add MD5Init, MD5Update, MD5Pad and
	MD5Final.

2005-07-26  Hector Garcia  <hector@debian.org>

	* Makefile (LIB_SRCS): Add 'md5c.c'.
	* md5c.c: Changed include to 'md5.h'.
	* md5.h: Replaced with the system header 'sys/md5.h'.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* fgetln.c (fgetln): Reindent, remove commented code and translate
	variable names to english.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* Versions: New file.
	* Makefile: Add versioned symbols support.



src/humanize_number.c
/*	$NetBSD: humanize_number.c,v 1.14 2008/04/28 20:22:59 martin Exp $	*/

/*
 * Copyright (c) 1997, 1998, 1999, 2002 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Jason R. Thorpe of the Numerical Aerospace Simulation Facility,
 * NASA Ames Research Center, by Luke Mewburn and by Tomas Svensson.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <assert.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <libutil.h>

int
humanize_number(char *buf, size_t len, int64_t bytes,
    const char *suffix, int scale, int flags)
{
	const char *prefixes, *sep;
	int	b, i, r, maxscale, s1, s2, sign;
	int64_t	divisor, max;
	size_t	baselen;

	assert(buf != NULL);
	assert(suffix != NULL);
	assert(scale >= 0);

	if (flags & HN_DIVISOR_1000) {
		/* SI for decimal multiplies */
		divisor = 1000;
		if (flags & HN_B)
			prefixes = "B\0k\0M\0G\0T\0P\0E";
		else
			prefixes = "\0\0k\0M\0G\0T\0P\0E";
	} else {
		/*
		 * binary multiplies
		 * XXX IEC 60027-2 recommends Ki, Mi, Gi...
		 */
		divisor = 1024;
		if (flags & HN_B)
			prefixes = "B\0K\0M\0G\0T\0P\0E";
		else
			prefixes = "\0\0K\0M\0G\0T\0P\0E";
	}

#define	SCALE2PREFIX(scale)	(&prefixes[(scale) << 1])
	maxscale = 7;

	if (scale >= maxscale &&
	    (scale & (HN_AUTOSCALE | HN_GETSCALE)) == 0)
		return (-1);

	if (buf == NULL || suffix == NULL)
		return (-1);

	if (len > 0)
		buf[0] = '\0';
	if (bytes < 0) {
		sign = -1;
		bytes *= -100;
		baselen = 3;		/* sign, digit, prefix */
	} else {
		sign = 1;
		bytes *= 100;
		baselen = 2;		/* digit, prefix */
	}
	if (flags & HN_NOSPACE)
		sep = "";
	else {
		sep = " ";
		baselen++;
	}
	baselen += strlen(suffix);

	/* Check if enough room for `x y' + suffix + `\0' */
	if (len < baselen + 1)
		return (-1);

	if (scale & (HN_AUTOSCALE | HN_GETSCALE)) {
		/* See if there is additional columns can be used. */
		for (max = 100, i = len - baselen; i-- > 0;)
			max *= 10;

		/*
		 * Divide the number until it fits the given column.
		 * If there will be an overflow by the rounding below,
		 * divide once more.
		 */
		for (i = 0; bytes >= max - 50 && i < maxscale; i++)
			bytes /= divisor;

		if (scale & HN_GETSCALE)
			return (i);
	} else
		for (i = 0; i < scale && i < maxscale; i++)
			bytes /= divisor;

	/* If a value <= 9.9 after rounding and ... */
	if (bytes < 995 && i > 0 && flags & HN_DECIMAL) {
		/* baselen + \0 + .N */
		if (len < baselen + 1 + 2)
			return (-1);
		b = ((int)bytes + 5) / 10;
		s1 = b / 10;
		s2 = b % 10;
		r = snprintf(buf, len, "%d%s%d%s%s%s",
		    sign * s1, localeconv()->decimal_point, s2,
		    sep, SCALE2PREFIX(i), suffix);
	} else
		r = snprintf(buf, len, "%" PRId64 "%s%s%s",
		    sign * ((bytes + 50) / 100),
		    sep, SCALE2PREFIX(i), suffix);

	return (r);
}
